File: app/src/main/java/com/todoroo/astrid/service/TaskDuplicator.java
package com.todoroo.astrid.service;

import static com.google.common.collect.Lists.transform;
import static com.todoroo.andlib.utility.DateUtilities.now;
import static com.todoroo.astrid.data.Task.NO_UUID;

import com.todoroo.astrid.dao.TaskDaoBlocking;
import com.todoroo.astrid.data.Task;
import com.todoroo.astrid.gcal.GCalHelper;
import java.util.ArrayList;
import java.util.List;
import javax.inject.Inject;
import org.tasks.LocalBroadcastManager;
import org.tasks.data.Alarm;
import org.tasks.data.AlarmDaoBlocking;
import org.tasks.data.CaldavDaoBlocking;
import org.tasks.data.CaldavTask;
import org.tasks.data.Geofence;
import org.tasks.data.GoogleTask;
import org.tasks.data.GoogleTaskDaoBlocking;
import org.tasks.data.LocationDaoBlocking;
import org.tasks.data.Tag;
import org.tasks.data.TagDaoBlocking;
import org.tasks.data.TagData;
import org.tasks.data.TagDataDaoBlocking;
import org.tasks.preferences.Preferences;

public class TaskDuplicator {

  private final GCalHelper gcalHelper;
  private final TaskDaoBlocking taskDao;
  private final TagDaoBlocking tagDao;
  private final TagDataDaoBlocking tagDataDao;
  private final GoogleTaskDaoBlocking googleTaskDao;
  private final CaldavDaoBlocking caldavDao;
  private final LocationDaoBlocking locationDao;
  private final AlarmDaoBlocking alarmDao;
  private final Preferences preferences;
  private final LocalBroadcastManager localBroadcastManager;

  @Inject
  public TaskDuplicator(
      GCalHelper gcalHelper,
      TaskDaoBlocking taskDao,
      LocalBroadcastManager localBroadcastManager,
      TagDaoBlocking tagDao,
      TagDataDaoBlocking tagDataDao,
      GoogleTaskDaoBlocking googleTaskDao,
      CaldavDaoBlocking caldavDao,
      LocationDaoBlocking locationDao,
      AlarmDaoBlocking alarmDao,
      Preferences preferences) {
    this.gcalHelper = gcalHelper;
    this.taskDao = taskDao;
    this.localBroadcastManager = localBroadcastManager;
    this.tagDao = tagDao;
    this.tagDataDao = tagDataDao;
    this.googleTaskDao = googleTaskDao;
    this.caldavDao = caldavDao;
    this.locationDao = locationDao;
    this.alarmDao = alarmDao;
    this.preferences = preferences;
  }

  public List<Task> duplicate(List<Long> taskIds) {
    List<Task> result = new ArrayList<>();
    for (Task task : taskDao.fetch(taskIds)) {
      result.add(clone(task));
    }
    localBroadcastManager.broadcastRefresh();
    return result;
  }

  private Task clone(Task clone) {
    long originalId = clone.getId();
    clone.setCreationDate(now());
    clone.setModificationDate(now());
    clone.setCompletionDate(0L);
    clone.setCalendarURI("");
    clone.setUuid(NO_UUID);
    clone.suppressSync();
    clone.suppressRefresh();
    taskDao.createNew(clone);

    List<TagData> tags = tagDataDao.getTagDataForTask(originalId);
    if (!tags.isEmpty()) {
      tagDao.insert(transform(tags, td -> new Tag(clone, td)));
    }

    GoogleTask googleTask = googleTaskDao.getByTaskId(originalId);
    boolean addToTop = preferences.addTasksToTop();
    if (googleTask != null) {
      googleTaskDao.insertAndShift(new GoogleTask(clone.getId(), googleTask.getListId()), addToTop);
    }

    CaldavTask caldavTask = caldavDao.getTask(originalId);
    if (caldavTask != null) {
      caldavDao.insert(clone, new CaldavTask(clone.getId(), caldavTask.getCalendar()), addToTop);
    }

    for (Geofence g : locationDao.getGeofencesForTask(originalId)) {
      locationDao.insert(
          new Geofence(clone.getId(), g.getPlace(), g.isArrival(), g.isDeparture(), g.getRadius()));
    }

    List<Alarm> alarms = alarmDao.getAlarms(originalId);
    if (!alarms.isEmpty()) {
      alarmDao.insert(transform(alarms, a -> new Alarm(clone.getId(), a.getTime())));
    }

    gcalHelper.createTaskEventIfEnabled(clone);

    taskDao.save(clone, null); // TODO: delete me

    return clone;
  }
}

File: app/src/main/java/com/todoroo/astrid/service/TaskDuplicator.kt
package com.todoroo.astrid.service

import com.todoroo.astrid.dao.TaskDao
import com.todoroo.astrid.gcal.GCalHelper
import org.tasks.LocalBroadcastManager
import org.tasks.data.dao.AlarmDao
import org.tasks.data.dao.CaldavDao
import org.tasks.data.dao.GoogleTaskDao
import org.tasks.data.dao.LocationDao
import org.tasks.data.dao.TagDao
import org.tasks.data.dao.TagDataDao
import org.tasks.data.dao.TaskAttachmentDao
import org.tasks.data.db.DbUtils.dbchunk
import org.tasks.data.entity.Alarm
import org.tasks.data.entity.Attachment
import org.tasks.data.entity.CaldavTask
import org.tasks.data.entity.Geofence
import org.tasks.data.entity.Tag
import org.tasks.data.entity.Task
import org.tasks.data.entity.Task.Companion.NO_ID
import org.tasks.preferences.Preferences
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import javax.inject.Inject

class TaskDuplicator @Inject constructor(
    private val gcalHelper: GCalHelper,
    private val taskDao: TaskDao,
    private val localBroadcastManager: LocalBroadcastManager,
    private val tagDao: TagDao,
    private val tagDataDao: TagDataDao,
    private val googleTaskDao: GoogleTaskDao,
    private val caldavDao: CaldavDao,
    private val locationDao: LocationDao,
    private val alarmDao: AlarmDao,
    private val preferences: Preferences,
    private val taskAttachmentDao: TaskAttachmentDao,
) {

    suspend fun duplicate(taskIds: List<Long>): List<Task> {
        return taskIds
            .dbchunk()
            .flatMap {
                it.minus(taskDao.getChildren(it).toSet())
            }
            .let { taskDao.fetch(it) }
            .filterNot { it.readOnly }
            .map { clone(it, it.parent) }
            .also { localBroadcastManager.broadcastRefresh() }
    }

    private suspend fun clone(task: Task, parentId: Long): Task {
        val clone = task.copy(
            id = NO_ID,
            creationDate = currentTimeMillis(),
            modificationDate = currentTimeMillis(),
            reminderLast = 0,
            completionDate = 0L,
            calendarURI = "",
            parent = parentId,
            remoteId = Task.NO_UUID,
        )
        clone.suppressSync()
        clone.suppressRefresh()
        val newId = taskDao.createNew(clone)
        val tags = tagDataDao.getTagDataForTask(task.id)
        if (tags.isNotEmpty()) {
            tagDao.insert(
                tags.map {
                    Tag(
                        task = clone.id,
                        taskUid = clone.uuid,
                        name = it.name,
                        tagUid = it.remoteId
                    )
                }
            )
        }
        val googleTask = googleTaskDao.getByTaskId(task.id)
        val caldavTask = caldavDao.getTask(task.id)
        if (googleTask != null) {
            googleTaskDao.insertAndShift(
                clone,
                CaldavTask(
                    task = clone.id,
                    calendar = googleTask.calendar,
                    remoteId = null
                ),
                preferences.addTasksToTop()
            )
        } else if (caldavTask != null) {
            val newDavTask = CaldavTask(
                task = clone.id,
                calendar = caldavTask.calendar
            )
            if (parentId != 0L) {
                val remoteParent = caldavDao.getRemoteIdForTask(parentId)
                newDavTask.remoteParent = remoteParent
            }
            caldavDao.insert(clone, newDavTask, preferences.addTasksToTop())
        }
        for (g in locationDao.getGeofencesForTask(task.id)) {
            locationDao.insert(
                    Geofence(
                        task = clone.id,
                        place = g.place,
                        isArrival = g.isArrival,
                        isDeparture = g.isDeparture,
                    )
            )
        }
        val alarms = alarmDao.getAlarms(task.id)
        if (alarms.isNotEmpty()) {
            alarmDao.insert(alarms.map { Alarm(task = clone.id, time = it.time, type = it.type) })
        }
        gcalHelper.createTaskEventIfEnabled(clone)
        taskDao.save(clone, null) // TODO: delete me
        taskAttachmentDao
            .getAttachmentsForTask(task.id)
            .map {
                Attachment(
                    task = clone.id,
                    fileId = it.fileId,
                    attachmentUid = it.attachmentUid
                )
            }
            .let { taskAttachmentDao.insert(it) }
        getDirectChildren(task.id).forEach { subtask ->
            clone(subtask, newId)
        }
        return clone
    }

    private suspend fun getDirectChildren(taskId: Long): List<Task> =
        taskDao
            .fetch(taskDao.getChildren(taskId))
            .filter { it.parent == taskId }
}